### PR TITLE
build(deps): bump datafusion 20240528

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Update datafusion to get substrait interval support. Also includes some upstream optimization to reduce clones.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: benchmarks/Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@ api.workspace = true
 arrow.workspace = true
 chrono.workspace = true
 clap.workspace = true
-client = { workspace = true, features = ["testing"] }
+# client = { workspace = true, features = ["testing"] }
```
Comment (evenyag) [inline benchmarks/Cargo.toml:15]: Remove it?

File: src/script/src/python/pyo3/builtins.rs [hunk 2]
```diff
@@ -100,14 +100,14 @@ pub(crate) fn greptime_builtins(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
...
-            median,
+            // median,
             approx_percentile_cont,
             array_agg,
             avg,
             correlation,
             count,
-            covariance,
-            covariance_pop,
+            // covariance,
```
Comment (evenyag) [inline src/script/src/python/pyo3/builtins.rs:109]: @discord9 Please check these changes. It removed some builtin functions.

File: tests/cases/standalone/common/types/decimal/decimal_aggregates.result [hunk 3]
```diff
@@ -3,9 +3,9 @@
 SELECT arrow_typeof(FIRST_VALUE('0.1'::DECIMAL(4,1)));
 
 +----------------------------------------+
-| arrow_typeof(FIRST_VALUE(Utf8("0.1"))) |
+| arrow_typeof(first_value(Utf8("0.1"))) |
 +----------------------------------------+
-| Decimal128(4, 1)                       |
+| Float64                                |
```
Comment (evenyag) [inline tests/cases/standalone/common/types/decimal/decimal_aggregates.result:8]: The type of a decimal is float64 😅

**Review by evenyag — COMMENTED** [review]
Great. Every time we upgrade datafusion needs 1k+ lines of code......

File: tests/cases/standalone/common/types/decimal/decimal_aggregates.result (hunk 3 above)
Comment (waynexia) [reply tests/cases/standalone/common/types/decimal/decimal_aggregates.result:8]: decimal is not one of the primitive `NUMERICS`. It got cast to the nearest float64

**Review by evenyag — APPROVED** [review]
Not sure whether the output of `first_value/last_value` is reasonable.

File: src/promql/src/extension_plan/empty_metric.rs [hunk 4]
```diff
@@ -156,19 +156,23 @@ impl UserDefinedLogicalNodeCore for EmptyMetric {
...
             start: self.start,
             end: self.end,
             interval: self.interval,
-            expr: if !expr.is_empty() {
-                Some(expr[0].clone())
+            expr: if !exprs.is_empty() {
+                Some(exprs[0].clone())
             } else {
                 None
             },
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/empty_metric.rs:172]:
```suggestion
            expr: exprs.into_iter().next(),
```

File: src/promql/src/extension_plan/histogram_fold.rs [hunk 5]
```diff
@@ -95,8 +95,12 @@ impl UserDefinedLogicalNodeCore for HistogramFold {
...
-        Self {
+    fn with_exprs_and_inputs(
+        &self,
+        _exprs: Vec<Expr>,
+        inputs: Vec<LogicalPlan>,
+    ) -> DataFusionResult<Self> {
+        Ok(Self {
             le_column: self.le_column.clone(),
             ts_column: self.ts_column.clone(),
             input: inputs[0].clone(),
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/histogram_fold.rs:106]:
```suggestion
            input: inputs.into_iter().next().unwrap(),
```

File: src/promql/src/extension_plan/instant_manipulate.rs [hunk 6]
```diff
@@ -83,18 +83,26 @@ impl UserDefinedLogicalNodeCore for InstantManipulate {
...
 
-        Self {
+        Ok(Self {
             start: self.start,
             end: self.end,
             lookback_delta: self.lookback_delta,
             interval: self.interval,
             time_index_column: self.time_index_column.clone(),
             field_column: self.field_column.clone(),
             input: inputs[0].clone(),
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/instant_manipulate.rs:104]:
```suggestion
            input: inputs.into_iter().next().unwrap(),
```

File: src/promql/src/extension_plan/normalize.rs [hunk 7]
```diff
@@ -81,15 +81,23 @@ impl UserDefinedLogicalNodeCore for SeriesNormalize {
...
+                "SeriesNormalize should have at least one input".to_string(),
+            ));
+        }
 
-        Self {
+        Ok(Self {
             offset: self.offset,
             time_index_column_name: self.time_index_column_name.clone(),
             need_filter_out_nan: self.need_filter_out_nan,
             input: inputs[0].clone(),
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/normalize.rs:99]:
```suggestion
            input: inputs.into_iter().next().unwrap()
```

File: src/promql/src/extension_plan/range_manipulate.rs [hunk 8]
```diff
@@ -243,7 +251,7 @@ impl UserDefinedLogicalNodeCore for RangeManipulate {
             field_columns: self.field_columns.clone(),
             input: inputs[0].clone(),
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/range_manipulate.rs:252]:
```suggestion
            input: inputs.into_iter().next().unwrap(),
```

File: src/promql/src/extension_plan/scalar_calculate.rs [hunk 9]
```diff
@@ -226,7 +234,7 @@ impl UserDefinedLogicalNodeCore for ScalarCalculate {
             field_column: self.field_column.clone(),
             input: inputs[0].clone(),
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/scalar_calculate.rs:235]:
```suggestion
            input: inputs.into_iter().next().unwrap(),
```

File: src/promql/src/extension_plan/union_distinct_on.rs [hunk 10]
```diff
@@ -135,18 +135,27 @@ impl UserDefinedLogicalNodeCore for UnionDistinctOn {
...
+        inputs: Vec<LogicalPlan>,
+    ) -> DataFusionResult<Self> {
+        if inputs.len() != 2 {
+            return Err(DataFusionError::Internal(
+                "UnionDistinctOn must have exactly 2 inputs".to_string(),
+            ));
+        }
 
         let left = inputs[0].clone();
         let right = inputs[1].clone();
```
Comment (fengjiachun) [inline src/promql/src/extension_plan/union_distinct_on.rs:150]:
```suggestion
        let mut inputs = inputs.into_iter();
        let left = inputs.next().unwrap();
        let right = inputs.next().unwrap();
```

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**